internal/service/dynamodb/global_secondary_index.go

```diff
@@ -0,0 +1,710 @@
+		}
+	}
+
+	for _, g := range table.GlobalSecondaryIndexes {
+		if res.Id() != aws.ToString(g.IndexArn) {
+			for _, ks := range g.KeySchema {
+				knownAttributes[aws.ToString(ks.AttributeName)] = knownAttributes[aws.ToString(ks.AttributeName)] + 1
+			}
+		}
+	}
+
+	ut := &dynamodb.UpdateTableInput{
+		TableName:            aws.String(res.Get("table").(string)),
+		AttributeDefinitions: []awstypes.AttributeDefinition{},
+		GlobalSecondaryIndexUpdates: []awstypes.GlobalSecondaryIndexUpdate{
+			{
+				Delete: &awstypes.DeleteGlobalSecondaryIndexAction{
+					IndexName: aws.String(res.Get(names.AttrName).(string)),
+				},
+			},
+		},
+	}
+
+	for _, ad := range table.AttributeDefinitions {
+		if knownAttributes[aws.ToString(ad.AttributeName)] > 0 {
+			ut.AttributeDefinitions = append(ut.AttributeDefinitions, ad)
+		}
+	}
+
+	if utRes, err := conn.UpdateTable(ctx, ut); err != nil {
+		return create.AppendDiagError(diags, names.DynamoDB, create.ErrActionUpdating, resNameGlobalSecondaryIndex, res.Id(), fmt.Errorf("creating GSI (%s): %w", res.Get(names.AttrName).(string), err))
+	} else {
+		for _, gsi := range utRes.TableDescription.GlobalSecondaryIndexes {
+			if aws.ToString(gsi.IndexName) == res.Get(names.AttrName).(string) {
+				if err := res.Set(names.AttrIndexARN, aws.ToString(gsi.IndexArn)); err != nil {
+					diags = append(diags, errs.NewErrorDiagnostic(fmt.Sprintf(`Unable to set field "%s"`, names.AttrIndexARN), fmt.Sprintf("Error: %v", err)))
+				}
+
+				res.SetId(aws.ToString(gsi.IndexArn))
+			}
+		}
+	}
+
+	if _, err = waitTableActive(ctx, conn, tableName, res.Timeout(schema.TimeoutUpdate)); err != nil {
+		return create.AppendDiagError(diags, names.DynamoDB, create.ErrActionWaitingForCreation, resNameGlobalSecondaryIndex, res.Id(), err)
+	}
+
+	if _, err := waitGSIDeleted(ctx, conn, tableName, res.Get(names.AttrName).(string), res.Timeout(schema.TimeoutUpdate)); err != nil {
+		return create.AppendDiagError(diags, names.DynamoDB, create.ErrActionWaitingForCreation, resNameGlobalSecondaryIndex, res.Id(), fmt.Errorf("GSI (%s): %w", res.Get(names.AttrName).(string), err))
+	}
+
+	if len(diags) > 0 {
+		return diags
+	}
+
+	return diags
+}
+
+func waitAllGSIActive(ctx context.Context, conn *dynamodb.Client, tableName string, timeout time.Duration) error {
+	stateConf := &retry.StateChangeConf{
+		Pending: enum.Slice(awstypes.TableStatusCreating, awstypes.TableStatusUpdating),
+		Target:  enum.Slice(awstypes.TableStatusActive),
+		Refresh: statusTable(ctx, conn, tableName),
+		Timeout: max(createTableTimeout, timeout),
+	}
+
+	outputRaw, err := stateConf.WaitForStateContext(ctx)
+	if err != nil {
+		return err
+	}
+
+	if output, ok := outputRaw.(*awstypes.TableDescription); ok {
+		for _, g := range output.GlobalSecondaryIndexes {
+			if g.IndexStatus != awstypes.IndexStatusActive {
+				if _, err := waitGSIActive(ctx, conn, tableName, aws.ToString(g.IndexName), timeout); err != nil {
+					return err
+				}
+			}
+		}
+	}
+
```

Comment:
Rather than calling `statusTable` (and recreating `waitTableActive`), the `Refresh` function could handle checking the status of all of the GSIs on the table with a single call to `DescribeTable`. Currently, `DescribeTable` is called at least once for `stateConf.WaitForStateContext` and at least once for each GSI.